Captains log, star date 2020...

impaktor opened the pull request:
## Introduction
Inspired by this [flight log](http://www.planetmic.com/orbit/fdl-file.htm) add-on for original Frontier, I've implemented something similar. This is highly work in progress. I've used the FlightLog that automatically logs departures from _stations_, and arrival/departure times into _systems_. I've also added a _custom_ log, that can be created by writing some fitting string, and pressing enter; and show all three in a Flight Log tab in InfoView.

~~Currently text can only be entered into the _custom_ logs, but I'll extend the other two logs (_system_, and _station_) so one can enter, as @nozmajner suggested: "this was a smelly station".~~

## Potential: Custom entries at start give backstory
I've played around with adding a custom log entry from the mainmenu, when starting a new game, so the first entry could be:

_"Just picked up the ship that uncle Nick left for me to inherrit. This rusty old bucket isn't much, but together we'll write our destiny in the stars, and it's my new home. Can't wait to undock!"_

*"Now I've cleared my debt with Wu, and sold all my belongings, and all I got was this lousy ship. By it's mine, and with it I can escape this god forsaken world and build a brighter future, forever onwards!"*


## Biggest hurdle
- Currently, it's to get the three logs (_system_, _station_, _custom_) to be printed in date-order, instead of printing each one in turn. I think the FlightLog method might have some method that sorts them, or combines them, but they have different table fields, so... need to think. 

- Multiline input. I've experimented with adding multiline text-input field, but the game crashes if I try to use it, so for now, it's just a dead commit.

- Also, this is a savebumping PR, since the updated FlightLog library will be incompatible with the previous, since I'm adding more fields (also log player's money), and a custom log.


## To do
- [x]  0 Allow re-editing old entries in custom logs
- [x]  1 Clear flightlog upon new game (bug of old?)
- [x]  2 Remove (only?) custom entries
-  3 Do we also want to be able to remove SYSTEM and STATION auto-logs as well?
- [x]  4 Add startup custom entry, backstory of how the ship ended up in the player's possession
- [x]  5 Edit/update Entries of SYSTEM and STATION auto-logs as well?
- [x]  6 ~~sort them on dates~~, not printing SYSTEM, STATION & CUSTOM logs separately. EDIT: Solved this by printing each log in a separate tab, thus #4893
- [x]  7 Better "Location" strings for CUSTOM, for if docked, landed, in orbit etc. 
- [x]  8 Document new methods in FlightLog
- [x]  9 Put strings in to translation system
- [x] 10 Use coloured text for some of the log-fields? E.g. Date, or similar?
- [x] 11 Better Log icon for InfoView tab
- [x] 12 Replace "Remove"-button with trashcan?
- [x] 13 Maybe/probably support multiline text box/input.
- [ ]  14 Save versioning of flightlog, to avoid save-bumping.
- ~~15. Make `textWrapped` include the secondary optional argument in LuaPiGui.cpp~~

## For future
These are things that might be intuitive next steps to add, but I don't plan to do them, so up for grabs:
- Export log to text-file?
- Filter logs
- Hyperlink systems/stations to map

## Current state
This is a screenshot of where I'm at. It works pretty well, as is, I think, ~~although only single line text-input, and~~ the tree logs are only date-sorted within each group (First it prints all _custom_ logs, then all _system_ logs, then all _station_ logs).

![2020-02-23-122512_1600x900_scrot](https://user-images.githubusercontent.com/619390/75111286-6b811380-5638-11ea-8068-7b49d6950087.png)

Closes #204

Web-eWorks commented:
@impaktor you could avoid bumping saves by simply serializing a Version field in the FlightLog object, and then checking if that field is nil and fixing up the log entries. It looks there aren't many changes which would need fixing up; you can simply display N/A if need be when there isn't data stored in a save file about how much money you had.

I'd like to retain backwards save compatibility as much as possible, and this really isn't a feature that's worth breaking saves over, in my opinion. It's a little more work, but it avoids invalidating saves.

impaktor commented:
- I've added more friendly versioning to the to-do list.

- The to-do list is numbered if anyone want to reference / give feedback to it. E.g. I don't know if we want to be able to remove the auto-generated log-entries (every system we've been to and every station we departed from).

- I've now combined the "Flight state" and "location" string, for _custom_ logs so, when the player creates a custom log entry, it will automatically generate a suitable string:

  - Location: Flying in space, relative to Sirius A
  - Location: Docked at Torvald's station, in orbit around Earth
  - Location: Docked at London station, on Earth
  - Location: Landed on Mars, <lat> <lon>

- I've updated the screenshot in the first post _(I've fixed the "::" seen in screenshot -- an artefact of me re-using strings from translation system)_

The-EG commented:
Opening a fairly new save (from master, a few weeks old at most) crashes:

[crashing_save.zip](https://github.com/pioneerspacesim/pioneer/files/4840554/crashing_save.zip)

impaktor commented:
OK, this was stalled since 4 months, because I didn't want to deal with the underlying three different data structures for logs, and how to sort them / weave them together (or re-write the whole thing). 

Solution: Tabs! And it also feels more feature rich this way, since now the player sees they got three logs, instead of just the one! For now, you can add text entry field to all three logs (custom log, auto-log system events, and auto-log station-events), but only the custom logs can be deleted fully. Maybe it's a feature, maybe it's a bug. I dunno.

## Things of interest for others to review

When starting a new game, there's some back story (and some hints):
 
![2020-06-27-152341_1600x900_scrot](https://user-images.githubusercontent.com/619390/85923339-710e9480-b88a-11ea-9a1b-9d7ec5c3df4e.png)

![2020-06-27-140320_1600x900_scrot](https://user-images.githubusercontent.com/619390/85923194-61428080-b889-11ea-8d12-a9d25292fdee.png)

![2020-06-27-152401_1600x900_scrot](https://user-images.githubusercontent.com/619390/85923340-766bdf00-b88a-11ea-8b74-ef3a85b1cfef.png)


## Remains?

This PR is now done, so I'm removing the WIP on it. Please hold off with merging this at least a few days. (I should probably squash the last commit into the first one, both are just adding documentation to lua functions)


## Future

There are many routes to develop this further, like integrating this with sector view, whith notes on systems, or filter and search function, or auto-logging buying/selling prices in station log, or/and kills, crimes, missions? (I suspect a re-write of the FlightLog module might be needed then). Anyway, I don't intend to do anything more on this, now, so I'll leave any future work to others.

robothauler commented:
![image](https://user-images.githubusercontent.com/10255455/85942405-5a6b4a80-b929-11ea-8e4c-4967c3c46f9b.png)

impaktor commented:
I've made the final tweaks on this now. I've changed it so station events are logged when docking, instead of un-docking, since I find that more intuitive.

Also, I realized when displaying the logs separately, station logs should also show which planet and system one docked in, or else one has to look in system log to see time stamp for different jumps.

Perhaps more (other) station events can be logged, like what the player buys, but I suspect that will require reworking flightlog library a bit.

Ready for final review.

impaktor commented:
@Web-eWorks if you want to hold this for review, then speak up, as I'd like to merge this the comming days.

I've noticed I have unused definition (`local orbiteer = ui.fonts.orbiteer`) that I could remove, but other than that, I'm happy with the code. (that's my self-review :smiley: )

impaktor commented:
I merged in master, and now, for some reason, this segfaults when launching pioneer, 
```
Thread 1 "pioneer" received signal SIGSEGV, Segmentation fault.
ImGui::BeginTabBar (str_id=0x7fffffffcd30 "mytabbar", flags=flags@entry=0) at /home/me/usr/src/pioneer/contrib/imgui/imgui_widgets.cpp:6407
6407	    if (window->SkipItems)
(gdb) bt
#0  ImGui::BeginTabBar (str_id=0x7fffffffcd30 "mytabbar", flags=flags@entry=0) at /home/me/usr/src/pioneer/contrib/imgui/imgui_widgets.cpp:6407
#1  0x00005555558887ce in l_pigui_begin_tab_bar (l=0x555555de8d70) at /usr/include/c++/10.1.0/bits/basic_string.h:2299
#2  0x00005555558c3756 in luaD_precall (L=L@entry=0x555555de8d70, func=<optimized out>, func@entry=0x5555566b4130, nresults=nresults@entry=1)
    at /home/me/usr/src/pioneer/contrib/lua/src/ldo.c:318
#3  0x00005555558d255d in luaV_execute (L=L@entry=0x555555de8d70) at /home/me/usr/src/pioneer/contrib/lua/src/lvm.c:709
#4  0x00005555558c3b28 in luaD_call (L=0x555555de8d70, func=<optimized out>, nResults=<optimized out>, allowyield=<optimized out>)
    at /home/me/usr/src/pioneer/contrib/lua/src/ldo.c:395
#5  0x00005555558c307c in luaD_rawrunprotected (L=L@entry=0x555555de8d70, f=f@entry=0x5555558ba0b0 <f_call>, ud=ud@entry=0x7fffffffcfe0)
    at /home/me/usr/src/pioneer/contrib/lua/src/ldo.c:131
#6  0x00005555558c3d5f in luaD_pcall (L=L@entry=0x555555de8d70, func=func@entry=0x5555558ba0b0 <f_call>, u=u@entry=0x7fffffffcfe0, old_top=112, ef=<optimized out>)
    at /home/me/usr/src/pioneer/contrib/lua/src/ldo.c:595
#7  0x00005555558bc937 in lua_pcallk (L=L@entry=0x555555de8d70, nargs=nargs@entry=0, nresults=nresults@entry=1, errfunc=errfunc@entry=6, ctx=ctx@entry=0, k=k@entry=0x0)
    at /home/me/usr/src/pioneer/contrib/lua/src/lapi.c:949
#8  0x00005555558b98b5 in pi_lua_dofile (l=l@entry=0x555555de8d70, code=..., nret=nret@entry=1) at /home/me/usr/src/pioneer/src/lua/core/Sandbox.cpp:209
#9  0x00005555558b5f82 in load_file_cached (L=0x555555de8d70, path="/pigui/modules/info-view/06-flightlog.lua", cacheIdx=5) at /home/me/usr/src/pioneer/src/SmartPtr.h:27
#10 0x00005555558b63c8 in lua_import_module (L=0x555555de8d70, moduleName="pigui.modules.info-view.06-flightlog") at /home/me/usr/src/pioneer/src/lua/core/Import.cpp:319
#11 0x00005555558b7c12 in pi_lua_import (popImported=true, importName="pigui.modules.info-view.06-flightlog", L=0x555555de8d70)
    at /home/me/usr/src/pioneer/src/lua/core/Import.cpp:407
#12 pi_lua_import_recursive (L=L@entry=0x555555de8d70, moduleName="pigui.modules") at /home/me/usr/src/pioneer/src/lua/core/Import.cpp:407
#13 0x0000555555866c7a in Lua::InitModules () at /usr/include/c++/10.1.0/bits/char_traits.h:322
#14 0x000055555561548a in std::function<void ()>::operator()() const (this=0x555555bcaa60) at /usr/include/c++/10.1.0/bits/std_function.h:617
#15 LoadStep::Update (this=<optimized out>, deltaTime=<optimized out>) at /home/me/usr/src/pioneer/src/Pi.cpp:614
#16 0x000055555596fc2f in Application::Run (this=0x555555b8fab0) at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:1324
#17 0x00005555555dfc70 in main (argc=<optimized out>, argv=<optimized out>) at /home/me/usr/src/pioneer/src/Pi.h:143
```
So I assume I've done something bad when I added the tab bar in #4893

Gliese852 commented:
@impaktor for some reason, the function `BeginTabBar` is triggered during module loading

The-EG commented:
At the end of 06-flightlog.lua, line 227, change `draw = drawLog()` to `draw = drawLog`. 

You want to assign the address of the function to `draw` not it's return value (which causes it to run).

impaktor commented:
@Gliese852 & @The-EG doh! yeah, thanks! Merging it now.

WKFO commented:
I tried this again recently and when I try to make an entry while in hyperspace, it does this:
![errorrr](https://user-images.githubusercontent.com/29180904/87412517-cd371f80-c5d1-11ea-8a18-ae5f7d2c1493.PNG)
I think this was fixed, though. Can anyone else give it a try?
